godel_process_execution/src/abb_blend_process_service.cpp

```diff
@@ -0,0 +1,235 @@
+
+  fp.flush();
+  return true;
+}
+
+godel_process_execution::AbbBlendProcessExecutionService::AbbBlendProcessExecutionService(const std::string& name, 
+                                                                                          const std::string& sim_name,
+                                                                                          const std::string& real_name,
+                                                                                          ros::NodeHandle& nh)
+  : name_(name)
+{
+  // Load Robot Specific Parameters
+  nh.param<bool>("J23_coupled", j23_coupled_, false);
+
+  // Create client services
+  sim_client_ = nh.serviceClient<simulator_service::SimulateTrajectory>(sim_name);
+
+  real_client_ = nh.serviceClient<abb_file_suite::ExecuteProgram>(real_name); // "execute_program"
+
+  // Advertise the primary blending service
+  server_ = nh.advertiseService<AbbBlendProcessExecutionService,
+                                godel_msgs::BlendProcessExecution::Request,
+                                godel_msgs::BlendProcessExecution::Response>
+            (name, &godel_process_execution::AbbBlendProcessExecutionService::executionCallback, this);
+}
+
+bool godel_process_execution::AbbBlendProcessExecutionService::executionCallback(godel_msgs::BlendProcessExecution::Request& req,
+                                                                                 godel_msgs::BlendProcessExecution::Response& res)
+{
+  if (req.simulate)
+  {
+    return simulateProcess(req);
+  }
+  else
+  {
+    return executeProcess(req);
+  }
+}
+
+bool godel_process_execution::AbbBlendProcessExecutionService::executeProcess(godel_msgs::BlendProcessExecution::Request req)
+{
+  trajectory_msgs::JointTrajectory aggregate_traj;
+  aggregate_traj = req.trajectory_approach;
+  appendTrajectory(aggregate_traj, req.trajectory_process);
+  appendTrajectory(aggregate_traj, req.trajectory_depart);
+
+  //ABB Rapid Emmiter
+  std::vector<rapid_emitter::TrajectoryPt> pts = toRapidTrajectory(aggregate_traj, j23_coupled_);
+
+  // RAPID process parameters
+  rapid_emitter::ProcessParams params;
+  params.spindle_speed = 1.0;
+  params.tcp_speed = 200;
+  params.wolf = false;
+  params.slide_force = 0.0;
+  params.output_name = "do_PIO_8";
+
+  // Calculate process start and end indexes
+  unsigned start_index = req.trajectory_approach.points.size();
+  unsigned stop_index = start_index + req.trajectory_process.points.size();
+
+  if (!writeRapidFile("/tmp/blend.mod", pts, start_index, stop_index, params))
+  {
+    ROS_WARN("Unable to generate RAPID motion file; Cannot execute process.");
+    return false;
+  }
+
+  // Call the ABB driver
+
+  abb_file_suite::ExecuteProgram srv;
+  srv.request.file_path = "/tmp/blend.mod";
+
+  if (!real_client_.call(srv))
+  {
+    ROS_WARN("Unable to upload blending process RAPID module to controller via FTP.");
+    return false;
+  }
+
+  if (!req.wait_for_execution)
+  {
+    return true;
```

Comment:
This one isn't an error condition. It's just saying if the user doesn't want to wait for the execution of the program, then return immediately.